core/src/test/java/com/datastax/oss/driver/api/core/cql/StatementBuilderTest.java

```diff
@@ -16,61 +16,95 @@
 package com.datastax.oss.driver.api.core.cql;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
 
+import java.nio.ByteBuffer;
+import java.util.Random;
 import org.junit.Test;
 
 public class StatementBuilderTest {
 
-  private static class NullStatementBuilder
-      extends StatementBuilder<NullStatementBuilder, SimpleStatement> {
+  private static class MockSimpleStatementBuilder
```

Comment:
Not sure we need this inner class? I just replaced all references to it with `SimpleStatementBuilder builder = new SimpleStatementBuilder("IRRELEVANT");` (or `builder = new SimpleStatementBuilder(template)`) and the tests passed.

Reply:
This came up in the PR for JAVA-2402 (which introduced the test in question).  The functionality we're testing here is implemented in StatementBuilder so my intent was to test that class directly.  Although it's not necessarily likely a subclass can always modify that behaviour; better to test as close to the functionality in question as possible.

Original comment should be at https://github.com/datastax/java-driver/pull/1330#discussion_r326422401

Reply:
IOW we are trying to test an abstract class, and that's why we had to introduce this concrete class. But we shouldn't be testing anything abstract in the first place, only its concrete implementations. Which are 3 currently:

1. BatchStatementBuilder
2. BoundStatementBuilder
3. SimpleStatementBuilder

So why don't you parameterize each test and make it run against an instance of each of the types above?

Reply:
I'm not sure I agree with the reasoning here.

There is a function implemented on the abstract super-class StatementBuilder.  That function should be tested just like any other function (so that we know it does the right thing if a sub-class doesn't override it).  We could test that by using an existing sub-class which doesn't (currently) override that method... but that test then is subject to silent changes in behaviour if that sub-class later overrides the method.

In order to avoid that concern, the unit test extends the abstract sub-class directly.

Testing all three concrete sub-classes _in this test_ seems like a bad idea for similar reasons; this test is intended to validate behaviours in the abstract super-class and the introduction of a new concrete sub-class requires an update to this unrelated test in a way that's not obvious.  Now, if you're arguing that we should have similar tests for each of the three concrete sub-classes _in addition to_ this test I could certainly go along with that.

I'm not opposed to testing the concrete sub-classes at all.  I just want to be very clear about the scope of each individual test.

Reply:
OK OK I don't feel strong about it. Creating a dummy child class to test a parent abstract class seems reasonable.